### DISCUSSION TITLE
Questions about what Nelua is

### DISCUSSION THREAD

**Discussion by JonesAndrew** [edubart/nelua-lang#10] [Q&A] [answered]
This project looks really cool, it's actually something I've been wondering if something like it exists for a while.

I just had some questions about the goals of the project just because some of the docs were a little unclear to me. Is this attempting to be able to take existing Lua code and then add type annotations and just have it work, or will it be more of a c language with a Lua flavor. What about features like coroutines?

And also just a few questions about the state of it now, is nelua entirely manual memory management or is there some type of gc or reference counting. From the looks of the generated code it seemed manual but just wanted to double check.

Anyways the project looks awesome and I wish you all good luck!

**Comment by edubart** [comment, marked as answer]
The near goal of the project is to have a C like language with Lua flavor. The long term goal is to have all possible Lua features that would not impact performance, including tables, its APIs, dynamic typing, coroutines, closures, garbage collecting and even metatables though all that would be used only depending in your code style. I want people that code Lua can code in Nelua with no much learning, they would only require to learn the type notations and other features in case of making efficient code, type safe code or meta programming.

However not all Lua features is in the goal, for example dynamic loading code is not a goal because Nelua doesn't have or will have an interpreter or a JIT. Thus functions like `load`, `loadbuffer`, `loadfile` will never be implemented. At the moment I don't plan to support exceptions too because it may interfere in many places and impact performance, so `error` function would trigger an unrecoverable error and `pcall` or `xpcall` would not be implemented.

Thus, right now you can use and think of Nelua as a C with Lua flavor, but is heading to a near Lua featureful language were Lua users could just use with no much learning and with low level C goodies for efficient code.

Right now the memory management can be manual or you can use a garbage collector. Reference counting memory management is planned and being worked on, some code for it already made in the compiler but is not finished and very experimental.

The language is evolving from like C features first and slowly the Lua features and APIs are being worked on.

**Comment by JonesAndrew** [comment]
Thanks, that clears stuff up for me. Thanks for the quick answer and good luck!
